services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithUnknownContentLengthHelper.java

```diff
@@ -160,6 +160,7 @@ public void onSubscribe(Subscription s) {
 
         @Override
         public void onNext(AsyncRequestBody asyncRequestBody) {
+            int currentPartNum = partNumber.incrementAndGet();
```

Comment:
Can you explain how we're using `partNumber` to me?

It looks like the first onNext, we increment partNumber to 1, then schedule the initiation of parts 1 and 2 (when the create MPU finishes).
The second onNext, we increment partNumber to 2 and then initiate another uploadPart request with part number 2?

Reply:
I see, it's because of our special handling of the first part. Would it be better to just increment partNumber right before we use it (not within whenComplete, but the else statement, before line 200, and initialize partNumber to 3 (using getAndIncrement instead)?

Nit: can we name it nextPartNumber (if getAndIncrement) or lastPartNumber (if incrementAndGet)?

Reply:
We schedule create MPU and the initiation of parts 1 and 2 in the *second* onNext, not the first because if onNext is only invoked once (meaning there's only one part), we use single object upload instead.